File: PolyScript/frameworks/elixir/lib/polyscript/framework.ex
# PolyScript Framework for Elixir
# CRUD × Modes Architecture: Zero-boilerplate CLI development
# 
# Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>

defmodule PolyScript.Framework do
  @moduledoc """
  PolyScript Framework for Elixir - provides CRUD × Modes CLI architecture
  """

  defmodule Nif do
    @moduledoc """
    Native Implemented Functions for libpolyscript integration
    """
    
    @on_load :init_nif
    
    def init_nif do
      # Try to load the NIF, but don't fail if it's not available
      try do
        nif_file = :filename.join(:code.priv_dir(:polyscript_framework), 'polyscript_nif')
        :erlang.load_nif(nif_file, 0)
      catch
        _type, _error ->
          IO.warn("libpolyscript NIF not found, using fallback implementations")
          :ok
      end
    end
    
    # NIF function declarations with fallback implementations
    def polyscript_can_mutate(_mode), do: :erlang.nif_error(:nif_not_loaded)
    def polyscript_should_validate(_mode), do: :erlang.nif_error(:nif_not_loaded)
    def polyscript_require_confirm(_mode, _operation), do: :erlang.nif_error(:nif_not_loaded)
    def polyscript_is_safe_mode(_mode), do: :erlang.nif_error(:nif_not_loaded)
    
    # Safe wrappers that catch NIF errors and fall back
    def safe_can_mutate(mode) do
      try do
        polyscript_can_mutate(mode_to_int(mode))
      catch
        :error, :nif_not_loaded -> mode == :live
        _type, _error -> mode == :live
      end
    end
    
    def safe_should_validate(mode) do
      try do
        polyscript_should_validate(mode_to_int(mode))
      catch
        :error, :nif_not_loaded -> mode == :sandbox
        _type, _error -> mode == :sandbox
      end
    end
    
    def safe_require_confirm(mode, operation, force) do
      if force do
        false
      else
        try do
          polyscript_require_confirm(mode_to_int(mode), operation_to_int(operation))
        catch
          :error, :nif_not_loaded -> 
            mode == :live and operation in [:update, :delete]
          _type, _error -> 
            mode == :live and operation in [:update, :delete]
        end
      end
    end
    
    def safe_is_safe_mode(mode) do
      try do
        polyscript_is_safe_mode(mode_to_int(mode))
      catch
        :error, :nif_not_loaded -> mode in [:simulate, :sandbox]
        _type, _error -> mode in [:simulate, :sandbox]
      end
    end
    
    # Convert Elixir atoms to C integers for NIF calls
    defp mode_to_int(:simulate), do: 0
    defp mode_to_int(:sandbox), do: 1
    defp mode_to_int(:live), do: 2
    
    defp operation_to_int(:create), do: 0
    defp operation_to_int(:read), do: 1
    defp operation_to_int(:update), do: 2
    defp operation_to_int(:delete), do: 3
  end

  defmodule Context do
    @moduledoc """
    Context passed to all tool operations
    """
    defstruct [
      :operation,
      :mode,
      :resource,
      :rebadged_as,
      :options,
      :verbose,
      :force,
      :json_output,
      :tool_name,
      output_data: %{
        "polyscript" => "1.0",
        "status" => "success",
        "data" => %{}
      },
      messages: []
    ]

    def can_mutate?(%__MODULE__{mode: mode}), do: Nif.safe_can_mutate(mode)

    def should_validate?(%__MODULE__{mode: mode}), do: Nif.safe_should_validate(mode)

    def require_confirm?(%__MODULE__{mode: mode, operation: operation, force: force}), 
      do: Nif.safe_require_confirm(mode, operation, force)

    def safe_mode?(%__MODULE__{mode: mode}), do: Nif.safe_is_safe_mode(mode)
  end

  @doc """
  Behaviour that all PolyScript tools must implement
  """
  @callback description() :: String.t()
  @callback create(resource :: String.t() | nil, options :: map(), context :: Context.t()) :: {:ok, map()} | {:error, String.t()}
  @callback read(resource :: String.t() | nil, options :: map(), context :: Context.t()) :: {:ok, map()} | {:error, String.t()}
  @callback update(resource :: String.t() | nil, options :: map(), context :: Context.t()) :: {:ok, map()} | {:error, String.t()}
  @callback delete(resource :: String.t() | nil, options :: map(), context :: Context.t()) :: {:ok, map()} | {:error, String.t()}

  @doc """
  Log a message to the context
  """
  def log(%Context{json_output: true, verbose: verbose} = context, message, level \\ :info) do
    msg = "[#{level}] #{message}"
    messages = [msg | context.messages]
    
    output_data = if verbose do
      Map.put(context.output_data, "messages", Enum.reverse(messages))
    else
      context.output_data
    end
    
    %{context | messages: messages, output_data: output_data}
  end

  def log(context, message, level) do
    case level do
      :error -> IO.puts(:stderr, "Error: #{message}")
      :warning -> IO.puts(:stderr, "Warning: #{message}")
      :info -> IO.puts(message)
      :debug -> if context.verbose, do: IO.puts("Debug: #{message}")
    end
    context
  end

  @doc """
  Output data from an operation
  """
  def output(%Context{json_output: true} = context, data, error \\ false) do
    output_data = if error do
      context.output_data
      |> Map.put("status", "error")
      |> Map.put("error", to_string(data))
    else
      case data do
        %{} = map ->
          Map.update!(context.output_data, "data", &Map.merge(&1, map))
        _ ->
          Map.put(context.output_data, "result", data)
      end
    end
    
    %{context | output_data: output_data}
  end

  def output(_context, data, error) do
    if error do
      IO.puts(:stderr, inspect(data))
    else
      Jason.encode!(data, pretty: true) |> IO.puts()
    end
  end

  @doc """
  Confirm an action with the user
  """
  def confirm(%Context{force: true}, _message), do: true
  
  def confirm(%Context{json_output: true} = context, message) do
    output(context, %{"confirmation_required" => message}, true)
    false
  end
  
  def confirm(_context, message) do
    response = IO.gets("#{message} [y/N]: ") |> String.trim() |> String.downcase()
    response in ["y", "yes"]
  end

  @doc """
  Execute an operation with mode wrapping
  """
  def execute_with_mode(tool_module, %Context{mode: mode} = context) do
    case mode do
      :simulate ->
        simulate_operation(tool_module, context)
      
      :sandbox ->
        sandbox_operation(tool_module, context)
      
      :live ->
        live_operation(tool_module, context)
    end
  end

  defp simulate_operation(tool_module, %Context{operation: :read} = context) do
    apply(tool_module, :read, [context.resource, context.options || %{}, context])
  end

  defp simulate_operation(_tool_module, context) do
    action_verb = case context.operation do
      :create -> "Would create"
      :update -> "Would update"
      :delete -> "Would delete"
      _ -> "Would read"
    end
    
    {:ok, %{
      "simulation" => true,
      "action" => "#{action_verb} #{context.resource || "resource"}",
      "options" => context.options || %{}
    }}
  end

  defp sandbox_operation(_tool_module, _context) do
    {:ok, %{
      "sandbox" => true,
      "validations" => %{
        "permissions" => "verified",
        "dependencies" => "available",
        "connectivity" => "established"
      },
      "ready" => true
    }}
  end

  defp live_operation(tool_module, context) do
    if Context.require_confirm?(context) do
      msg = "Are you sure you want to #{context.operation} #{context.resource || "resource"}?"
      if !confirm(context, msg) do
        {:ok, %{"cancelled" => true}}
      else
        execute_operation(tool_module, context)
      end
    else
      execute_operation(tool_module, context)
    end
  end

  defp execute_operation(tool_module, context) do
    apply(tool_module, context.operation, [context.resource, context.options || %{}, context])
  end

  @doc """
  Finalize and output the context
  """
  def finalize_output(%Context{json_output: true} = context) do
    output_data = context.output_data
    |> Map.put("operation", to_string(context.operation))
    |> Map.put("mode", to_string(context.mode))
    |> Map.put("tool", context.tool_name)
    |> maybe_add_field("resource", context.resource)
    |> maybe_add_field("rebadged_as", context.rebadged_as)
    
    Jason.encode!(output_data, pretty: true) |> IO.puts()
  end

  def finalize_output(_context), do: :ok

  defp maybe_add_field(map, _key, nil), do: map
  defp maybe_add_field(map, key, value), do: Map.put(map, key, value)

  @doc """
  Parse command line arguments for a tool
  """
  def parse_args(argv, tool_module) do
    {opts, args, _} = OptionParser.parse(argv,
      switches: [
        mode: :string,
        verbose: :boolean,
        force: :boolean,
        json: :boolean,
        discover: :boolean,
        help: :boolean
      ],
      aliases: [
        m: :mode,
        v: :verbose,
        f: :force,
        h: :help
      ]
    )

    cond do
      opts[:help] -> {:help}
      opts[:discover] -> {:discover}
      length(args) == 0 -> {:error, "No operation specified"}
      true ->
        [operation | rest] = args
        resource = List.first(rest)
        
        case parse_operation(operation) do
          {:ok, op} ->
            mode = parse_mode(opts[:mode] || "live")
            context = %Context{
              operation: op,
              mode: mode,
              resource: resource,
              options: opts,
              verbose: opts[:verbose] || false,
              force: opts[:force] || false,
              json_output: opts[:json] || false,
              tool_name: tool_module |> Module.split() |> List.last()
            }
            {:ok, context}
          
          :error ->
            {:error, "Unknown operation: #{operation}"}
        end
    end
  end

  defp parse_operation("create"), do: {:ok, :create}
  defp parse_operation("read"), do: {:ok, :read}
  defp parse_operation("list"), do: {:ok, :read}
  defp parse_operation("update"), do: {:ok, :update}
  defp parse_operation("delete"), do: {:ok, :delete}
  defp parse_operation(_), do: :error

  defp parse_mode("simulate"), do: :simulate
  defp parse_mode("sandbox"), do: :sandbox
  defp parse_mode("live"), do: :live
  defp parse_mode(_), do: :live

  @doc """
  Show discovery information for a tool
  """
  def show_discovery(tool_module) do
    discovery = %{
      "polyscript" => "1.0",
      "tool" => tool_module |> Module.split() |> List.last(),
      "operations" => ["create", "read", "update", "delete"],
      "modes" => ["simulate", "sandbox", "live"]
    }
    
    Jason.encode!(discovery, pretty: true) |> IO.puts()
  end

  @doc """
  Show help for a tool
  """
  def show_help(tool_module) do
    tool_name = tool_module |> Module.split() |> List.last() |> String.downcase()
    description = apply(tool_module, :description, [])
    
    IO.puts("""
    #{tool_name} - #{description}
    
    Usage:
      #{tool_name} <operation> [resource] [options]
    
    Operations:
      create    Create new resources
      read      Read/query resources
      list      List resources (alias for read)
      update    Update existing resources
      delete    Delete resources
    
    Options:
      -m, --mode      Execution mode (simulate, sandbox, live) [default: live]
      -v, --verbose   Enable verbose output
      -f, --force     Skip confirmation prompts
      --json          Output in JSON format
      --discover      Show tool capabilities
      -h, --help      Show this help
    
    Examples:
      #{tool_name} create myfile.txt --mode simulate
      #{tool_name} read --json
      #{tool_name} delete old-data --force
    """)
  end

  @doc """
  Main entry point for running a PolyScript tool
  """
  def run(argv, tool_module) do
    case parse_args(argv, tool_module) do
      {:help} ->
        show_help(tool_module)
        System.halt(0)
      
      {:discover} ->
        show_discovery(tool_module)
        System.halt(0)
      
      {:error, msg} ->
        IO.puts(:stderr, "Error: #{msg}")
        System.halt(1)
      
      {:ok, context} ->
        try do
          context = log(context, "Executing #{context.operation} operation in #{context.mode} mode", :debug)
          
          case execute_with_mode(tool_module, context) do
            {:ok, result} ->
              context = output(context, result)
              finalize_output(context)
              System.halt(0)
            
            {:error, error} ->
              context = output(context, error, true)
              finalize_output(context)
              System.halt(1)
          end
        rescue
          e ->
            error_msg = Exception.message(e)
            context = output(context, "Unhandled error: #{error_msg}", true)
            
            if context.verbose do
              IO.puts(:stderr, Exception.format(:error, e, __STACKTRACE__))
            end
            
            finalize_output(context)
            System.halt(1)
        end
    end
  end
end

File: PolyScript/frameworks/scala/src/main/scala/polyscript/Framework.scala
/*
 * PolyScript Framework for Scala
 * CRUD × Modes Architecture: Zero-boilerplate CLI development
 * 
 * Author: Mathew Burkitt, Dimension Technologies <mathew.burkitt@ditech.ai>
 */

package polyscript

import scopt.OParser
import play.api.libs.json._
import java.time.Instant
import scala.util.{Try, Success, Failure}
import com.sun.jna.{Library, Native}
import java.io.File

object Framework {
  
  // JNA interface for libpolyscript
  trait LibPolyScript extends Library {
    def polyscript_can_mutate(mode: Int): Boolean
    def polyscript_should_validate(mode: Int): Boolean  
    def polyscript_require_confirm(mode: Int, operation: Int): Boolean
    def polyscript_is_safe_mode(mode: Int): Boolean
  }
  
  // FFI wrapper with graceful fallback
  object PolyScriptFFI {
    private var libpolyscript: Option[LibPolyScript] = None
    
    // Try to load libpolyscript from various locations
    private val libPaths = Seq(
      "../../libpolyscript/build/libpolyscript",
      "libpolyscript"  // System-wide installation
    )
    
    // Initialize FFI library
    try {
      var loaded = false
      for (path <- libPaths if !loaded) {
        try {
          libpolyscript = Some(Native.load(path, classOf[LibPolyScript]))
          loaded = true
        } catch {
          case _: UnsatisfiedLinkError => // Try next path
        }
      }
      if (!loaded) {
        System.err.println("Warning: libpolyscript not found, using fallback implementations")
      }
    } catch {
      case e: Exception =>
        System.err.println(s"Warning: Failed to load libpolyscript: ${e.getMessage}")
    }
    
    // Convert Scala types to C integers for FFI calls
    private def operationToInt(op: PolyScriptOperation): Int = op match {
      case Create => 0
      case Read => 1
      case Update => 2
      case Delete => 3
    }
    
    private def modeToInt(mode: PolyScriptMode): Int = mode match {
      case Simulate => 0
      case Sandbox => 1
      case Live => 2
    }
    
    // Safe FFI wrappers with fallback
    def canMutate(mode: PolyScriptMode): Boolean = {
      libpolyscript.map(_.polyscript_can_mutate(modeToInt(mode))).getOrElse(mode == Live)
    }
    
    def shouldValidate(mode: PolyScriptMode): Boolean = {
      libpolyscript.map(_.polyscript_should_validate(modeToInt(mode))).getOrElse(mode == Sandbox)
    }
    
    def requireConfirm(mode: PolyScriptMode, operation: PolyScriptOperation, force: Boolean): Boolean = {
      if (force) false
      else libpolyscript.map(_.polyscript_require_confirm(modeToInt(mode), operationToInt(operation)))
        .getOrElse(mode == Live && (operation == Update || operation == Delete))
    }
    
    def isSafeMode(mode: PolyScriptMode): Boolean = {
      libpolyscript.map(_.polyscript_is_safe_mode(modeToInt(mode))).getOrElse(mode == Simulate || mode == Sandbox)
    }
  }
  
  // CRUD Operations
  sealed trait PolyScriptOperation
  case object Create extends PolyScriptOperation
  case object Read extends PolyScriptOperation
  case object Update extends PolyScriptOperation
  case object Delete extends PolyScriptOperation
  
  // Execution Modes
  sealed trait PolyScriptMode
  case object Simulate extends PolyScriptMode
  case object Sandbox extends PolyScriptMode
  case object Live extends PolyScriptMode
  
  // Context for operations
  case class PolyScriptContext(
    operation: PolyScriptOperation,
    mode: PolyScriptMode,
    resource: Option[String] = None,
    rebadgedAs: Option[String] = None,
    options: Map[String, Any] = Map.empty,
    verbose: Boolean = false,
    force: Boolean = false,
    jsonOutput: Boolean = false,
    toolName: String = "",
    var outputData: Map[String, JsValue] = Map(
      "polyscript" -> JsString("1.0"),
      "status" -> JsString("success"),
      "data" -> JsObject(Seq.empty)
    ),
    var messages: List[String] = List.empty
  ) {
    def canMutate: Boolean = PolyScriptFFI.canMutate(mode)
    def shouldValidate: Boolean = PolyScriptFFI.shouldValidate(mode)
    def requireConfirm: Boolean = PolyScriptFFI.requireConfirm(mode, operation, force)
    def isSafeMode: Boolean = PolyScriptFFI.isSafeMode(mode)
    
    def log(message: String, level: String = "info"): Unit = {
      if (jsonOutput) {
        messages = messages :+ s"[$level] $message"
        if (verbose) {
          outputData = outputData + ("messages" -> JsArray(messages.map(JsString)))
        }
      } else {
        level match {
          case "error" => Console.err.println(s"Error: $message")
          case "warning" => Console.err.println(s"Warning: $message")
          case "info" => println(message)
          case "debug" if verbose => println(s"Debug: $message")
          case _ =>
        }
      }
    }
    
    def output(data: JsValue, error: Boolean = false): Unit = {
      if (jsonOutput) {
        if (error) {
          outputData = outputData + ("status" -> JsString("error"))
          outputData = outputData + ("error" -> data)
        } else {
          outputData = outputData + ("result" -> data)
        }
      } else {
        if (error) {
          Console.err.println(Json.prettyPrint(data))
        } else {
          println(Json.prettyPrint(data))
        }
      }
    }
    
    def confirm(message: String): Boolean = {
      if (force) return true
      
      if (jsonOutput) {
        output(Json.obj("confirmation_required" -> message), error = true)
        false
      } else {
        print(s"$message [y/N]: ")
        val response = scala.io.StdIn.readLine().toLowerCase
        response == "y" || response == "yes"
      }
    }
    
    def finalizeOutput(): Unit = {
      outputData = outputData ++ Map(
        "operation" -> JsString(operation.toString.toLowerCase),
        "mode" -> JsString(mode.toString.toLowerCase),
        "tool" -> JsString(toolName)
      )
      
      resource.foreach(r => outputData = outputData + ("resource" -> JsString(r)))
      rebadgedAs.foreach(r => outputData = outputData + ("rebadged_as" -> JsString(r)))
      
      if (jsonOutput) {
        println(Json.prettyPrint(JsObject(outputData)))
      }
    }
  }
  
  // Tool trait that all PolyScript tools must implement
  trait PolyScriptTool {
    def description: String
    def create(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue]
    def read(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue]
    def update(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue]
    def delete(resource: Option[String], options: Map[String, Any], context: PolyScriptContext): Try[JsValue]
  }
  
  // Command line configuration
  case class Config(
    operation: Option[String] = None,
    resource: Option[String] = None,
    mode: PolyScriptMode = Live,
    verbose: Boolean = false,
    force: Boolean = false,
    json: Boolean = false,
    discover: Boolean = false
  )
  
  // Execute operation with mode wrapping
  def executeWithMode(tool: PolyScriptTool, context: PolyScriptContext): Try[JsValue] = {
    context.mode match {
      case Simulate =>
        context.log(s"Simulating ${context.operation} operation", "debug")
        
        if (context.operation == Read) {
          tool.read(context.resource, context.options, context)
        } else {
          val actionVerb = context.operation match {
            case Create => "Would create"
            case Update => "Would update"
            case Delete => "Would delete"
            case _ => "Would read"
          }
          
          Success(Json.obj(
            "simulation" -> true,
            "action" -> s"$actionVerb ${context.resource.getOrElse("resource")}",
            "options" -> JsObject(context.options.map { case (k, v) => k -> JsString(v.toString) })
          ))
        }
        
      case Sandbox =>
        context.log(s"Testing prerequisites for ${context.operation}", "debug")
        
        Success(Json.obj(
          "sandbox" -> true,
          "validations" -> Json.obj(
            "permissions" -> "verified",
            "dependencies" -> "available",
            "connectivity" -> "established"
          ),
          "ready" -> true
        ))
        
      case Live =>
        context.log(s"Executing ${context.operation} operation", "debug")
        
        if (context.requireConfirm) {
          val msg = s"Are you sure you want to ${context.operation.toString.toLowerCase} ${context.resource.getOrElse("resource")}?"
          if (!context.confirm(msg)) {
            context.outputData = context.outputData + ("status" -> JsString("cancelled"))
            return Success(Json.obj("cancelled" -> true))
          }
        }
        
        context.operation match {
          case Create => tool.create(context.resource, context.options, context)
          case Read => tool.read(context.resource, context.options, context)
          case Update => tool.update(context.resource, context.options, context)
          case Delete => tool.delete(context.resource, context.options, context)
        }
    }
  }
  
  // Show discovery information
  def showDiscovery(toolName: String): Unit = {
    val discovery = Json.obj(
      "polyscript" -> "1.0",
      "tool" -> toolName,
      "operations" -> Json.arr("create", "read", "update", "delete"),
      "modes" -> Json.arr("simulate", "sandbox", "live")
    )
    println(Json.prettyPrint(discovery))
  }
  
  // Main entry point for running a tool
  def run[T <: PolyScriptTool](args: Array[String], tool: T): Unit = {
    val toolName = tool.getClass.getSimpleName.replace("$", "")
    
    val builder = OParser.builder[Config]
    val parser = {
      import builder._
      OParser.sequence(
        programName(toolName.toLowerCase),
        head(toolName, "1.0.0"),
        help('h', "help").text("Show this help message"),
        opt[Unit]("discover")
          .action((_, c) => c.copy(discover = true))
          .text("Show tool capabilities"),
        opt[String]('m', "mode")
          .action((x, c) => c.copy(mode = x.toLowerCase match {
            case "simulate" => Simulate
            case "sandbox" => Sandbox
            case _ => Live
          }))
          .text("Execution mode (simulate, sandbox, live)"),
        opt[Unit]('v', "verbose")
          .action((_, c) => c.copy(verbose = true))
          .text("Enable verbose output"),
        opt[Unit]('f', "force")
          .action((_, c) => c.copy(force = true))
          .text("Skip confirmation prompts"),
        opt[Unit]("json")
          .action((_, c) => c.copy(json = true))
          .text("Output in JSON format"),
        cmd("create")
          .action((_, c) => c.copy(operation = Some("create")))
          .text("Create new resources")
          .children(
            arg[String]("<resource>").optional()
              .action((x, c) => c.copy(resource = Some(x)))
              .text("Resource to create")
          ),
        cmd("read")
          .action((_, c) => c.copy(operation = Some("read")))
          .text("Read/query resources")
          .children(
            arg[String]("<resource>").optional()
              .action((x, c) => c.copy(resource = Some(x)))
              .text("Resource to read")
          ),
        cmd("list")
          .action((_, c) => c.copy(operation = Some("read")))
          .text("List resources (alias for read)"),
        cmd("update")
          .action((_, c) => c.copy(operation = Some("update")))
          .text("Update existing resources")
          .children(
            arg[String]("<resource>").optional()
              .action((x, c) => c.copy(resource = Some(x)))
              .text("Resource to update")
          ),
        cmd("delete")
          .action((_, c) => c.copy(operation = Some("delete")))
          .text("Delete resources")
          .children(
            arg[String]("<resource>").optional()
              .action((x, c) => c.copy(resource = Some(x)))
              .text("Resource to delete")
          ),
        checkConfig { c =>
          if (c.discover) success
          else if (c.operation.isEmpty) failure("No operation specified")
          else success
        }
      )
    }
    
    OParser.parse(parser, args, Config()) match {
      case Some(config) =>
        if (config.discover) {
          showDiscovery(toolName)
          System.exit(0)
        }
        
        val operation = config.operation.get match {
          case "create" => Create
          case "read" => Read
          case "update" => Update
          case "delete" => Delete
        }
        
        val context = PolyScriptContext(
          operation = operation,
          mode = config.mode,
          resource = config.resource,
          verbose = config.verbose,
          force = config.force,
          jsonOutput = config.json,
          toolName = toolName
        )
        
        try {
          context.log(s"Executing ${context.operation} operation in ${context.mode} mode", "debug")
          
          executeWithMode(tool, context) match {
            case Success(result) =>
              context.output(result)
              context.finalizeOutput()
              System.exit(0)
              
            case Failure(exception) =>
              context.output(JsString(s"Error: ${exception.getMessage}"), error = true)
              if (context.verbose) {
                exception.printStackTrace()
              }
              context.finalizeOutput()
              System.exit(1)
          }
        } catch {
          case e: Exception =>
            context.output(JsString(s"Unhandled error: ${e.getMessage}"), error = true)
            if (context.verbose) {
              e.printStackTrace()
            }
            context.finalizeOutput()
            System.exit(1)
        }
        
      case _ =>
        System.exit(1)
    }
  }
}